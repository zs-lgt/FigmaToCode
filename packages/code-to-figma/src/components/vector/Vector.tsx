import * as React from 'react';
import {
    DefaultShapeProps,
    VectorNodeProps,
    CornerProps,
    StyleOf,
    InstanceItemProps,
    SelectionEventProps
} from '../../types';
import {
    LayoutStyleProperties,
    transformLayoutStyleProperties
} from '../../styleTransformers/transformLayoutStyleProperties';
import { useYogaLayout } from '../../hooks/useYogaLayout';
import { transformBlendProperties, BlendStyleProperties } from '../../styleTransformers/transformBlendProperties';
import {
    transformGeometryStyleProperties,
    GeometryStyleProperties
} from '../../styleTransformers/transformGeometryStyleProperties';
import { YogaStyleProperties } from '../../yoga/YogaStyleProperties';
import { StyleSheet } from '../..';
import { useSelectionChange } from '../../hooks/useSelectionChange';
import { transformAutoLayoutToYoga } from '../../styleTransformers/transformAutoLayoutToYoga';
import { OnLayoutHandlerProps, useOnLayoutHandler } from '../../hooks/useOnLayoutHandler';
import { useImageHash } from '../../hooks/useImageHash';
import { useNodeIdCallback } from '../../hooks/useNodeIdCallback';

export interface VectorProps
    extends VectorNodeProps,
        DefaultShapeProps,
        CornerProps,
        InstanceItemProps,
        SelectionEventProps,
        OnLayoutHandlerProps {
    style?: StyleOf<YogaStyleProperties & LayoutStyleProperties & BlendStyleProperties & GeometryStyleProperties>;
    children?: undefined;
}

const Vector: React.FC<VectorProps> = props => {
    const nodeRef = React.useRef();

    useSelectionChange(nodeRef, props);
    useNodeIdCallback(nodeRef, props.onNodeId);

    const style = { ...StyleSheet.flatten(props.style), ...transformAutoLayoutToYoga(props) };

    const imageHash = useImageHash(style.backgroundImage);

    const vectorProps = {
        ...transformLayoutStyleProperties(style),
        ...transformBlendProperties(style),
        ...transformGeometryStyleProperties('fills', style, imageHash),
        ...props,
        style
    };
    const yogaProps = useYogaLayout({ nodeRef, ...vectorProps });
    useOnLayoutHandler(yogaProps, props);
    return <vector {...vectorProps} {...yogaProps} innerRef={nodeRef} />;
};

export { Vector };
