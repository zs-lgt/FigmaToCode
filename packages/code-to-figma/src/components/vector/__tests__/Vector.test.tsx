import * as React from 'react';
const renderer = require('react-test-renderer');
import { Vector } from '../Vector';
import { createFigma } from 'figma-api-stub';

describe('<Vector />', () => {
    beforeEach(() => {
        // @ts-ignore
        global.figma = createFigma({
            simulateErrors: true
        });
    });

    it('Vector without props', () => {
        const tree = renderer.create(<Vector />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Vector that draws a triangle', () => {
        const tree = renderer
            .create(
                <Vector
                    vectorNetwork={{
                        // The vertices of the triangle
                        vertices: [{ x: 0, y: 100 }, { x: 100, y: 100 }, { x: 50, y: 0 }],

                        // The edges of the triangle.
                        segments: [
                            {
                                start: 0,
                                end: 1
                            },
                            {
                                start: 1,
                                end: 2
                            },
                            {
                                start: 2,
                                end: 0
                            }
                        ],

                        // The loop that forms the triangle. Each loop is a
                        // sequence of indices into the segments array.
                        regions: [{ windingRule: 'NONZERO', loops: [[0, 1, 2]] }]
                    }}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
