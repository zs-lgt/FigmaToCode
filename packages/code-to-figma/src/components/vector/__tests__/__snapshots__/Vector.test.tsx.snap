// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Vector /> Vector that draws a triangle 1`] = `
<vector
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={Object {}}
  vectorNetwork={
    Object {
      "regions": Array [
        Object {
          "loops": Array [
            Array [
              0,
              1,
              2,
            ],
          ],
          "windingRule": "NONZERO",
        },
      ],
      "segments": Array [
        Object {
          "end": 1,
          "start": 0,
        },
        Object {
          "end": 2,
          "start": 1,
        },
        Object {
          "end": 0,
          "start": 2,
        },
      ],
      "vertices": Array [
        Object {
          "x": 0,
          "y": 100,
        },
        Object {
          "x": 100,
          "y": 100,
        },
        Object {
          "x": 50,
          "y": 0,
        },
      ],
    }
  }
/>
`;

exports[`<Vector /> Vector without props 1`] = `
<vector
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={Object {}}
/>
`;
